src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesUtils.java

```diff
@@ -216,6 +266,209 @@ public void processElement(final ProcessContext context) throws Exception {
                 });
     }
 
+    // This transform takes marked reads and generates the metrics from those tests.
+    public static class GenerateMetricsTransform extends PTransform<PCollection<GATKRead>, PCollection<KV<String, DuplicationMetrics>>> {
+        private static final long serialVersionUID = 1l;
+
+        /**
+         *  The header for all reads processed.
+        */
+        private final PCollectionView<SAMFileHeader> header;
+
+        public GenerateMetricsTransform(final PCollectionView<SAMFileHeader> header) {
+            this.header = header;
+        }
+
+        @Override
+        public PCollection<KV<String, DuplicationMetrics>> apply(final PCollection<GATKRead> preads) {
+            return preads
+                .apply(generateMetricsByLibrary())
+                .setCoder(KvCoder.of(StringUtf8Coder.of(), new DuplicationMetricsCoder()))
+                // Combine metrics for reads from the same library.
+                .apply(Combine.<String, DuplicationMetrics>perKey(new CombineMetricsFn()))
+                // For each library, finalize the metrics by computing derived metrics and dividing paired counters
+                // by 2.
+                .apply(finalizeMetrics());
+        }
+
+        private PTransform<PCollection<? extends GATKRead>, PCollection<KV<String, DuplicationMetrics>>> generateMetricsByLibrary() {
+            return ParDo
+                .named("generate metrics by library")
+                .withSideInputs(header)
+                .of(new DoFn<GATKRead, KV<String, DuplicationMetrics>>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            final GATKRead record = context.element();
+                            if (record.isSecondaryAlignment() || record.isSupplementaryAlignment()) {
+                              return;
+                            }
+                            final SAMFileHeader h = context.sideInput(header);
+                            final String library = LibraryIdGenerator.getLibraryName(h, record.getReadGroup());
+                            DuplicationMetrics metrics = new DuplicationMetrics();
+                            metrics.LIBRARY = library;
+                            // These need to be initialized so that the DuplicationMetrics can be serialized safely.
+                            metrics.PERCENT_DUPLICATION = 0.0;
```

Comment:
this initialization code should move into the constructor if it's necessary that shouldn't break anything elsewhere I hope?  

I think it would be avoided by making DuplicationMetrics serializable and using a serializable coder instead of the custom one though